File: src/template/haskell/bfs.hs
import qualified Data.Sequence as Seq
import Data.Foldable (toList)

bfs :: Int -> [[Int]] -> [Int]
bfs start adj = go (Seq.singleton start) (replicate n maxBound) where
  n = length adj
  go Seq.Empty dist = dist
  go (u Seq.:<| q) dist =
    let (dist', q') = foldl (\(d, qs) v ->
          if d !! v == maxBound then
            (take v d ++ [d !! u + 1] ++ drop (v+1) d, qs Seq.|> v)
          else (d, qs)
        ) (dist, q) (adj !! u)
    in go q' dist'

File: src/template/rust/bfs.rs
use std::collections::VecDeque;

pub fn bfs(start: usize, adj: &Vec<Vec<usize>>) -> Vec<usize> {
    let n = adj.len();
    let mut dist = vec![usize::MAX; n];
    let mut q = VecDeque::new();
    dist[start] = 0;
    q.push_back(start);
    while let Some(u) = q.pop_front() {
        for &v in &adj[u] {
            if dist[v] == usize::MAX {
                dist[v] = dist[u] + 1;
                q.push_back(v);
            }
        }
    }
    dist
}